### PR TITLE
Adds a DFS and a BFS solution for "The Maze"

### PR DESCRIPTION
**Pull request by altayhunoglu**
Resolves https://github.com/spiralgo/algorithms/issues/48
 
 This solution is very important to grasp DFS and it is a kind of basis for [The Maze II](https://github.com/spiralgo/algorithms/issues/56) and [The Maze III](https://github.com/spiralgo/algorithms/issues/231).
 
 It is also a little bit different than the usual DFS problems as you cannot stop and change direction until you encounter a wall or the destination.
 
 It is important to note that, you have to decide if DFS **can pass** over the **next** cell, not the **current** cell.
 
 The most straightforward cases are noted in the problem statement (https://github.com/spiralgo/algorithms/issues/48) for the first and second examples.
 
 But, I think there is another case that is more veiled and pretty hard to see at the first glance.
 
 I tried to draw the mentioned case:
 
IMG-1626

 In the above-mentioned case; you must pass at the first encounter but you have to stop after another turn of DFS (and in a different direction.)

### CODE REVIEW

**Comment by altayhunoglu** [comment]
image

**Comment by ErdemT09** [comment]
Why is the helper method called "dps" and not "dfs"?

**Comment by altayhunoglu** [comment]
> Why is the helper method called "dps" and not "dfs"?
> Does it stand for something like "depth prioritized search"?

No no. It is just a typo. Thanks for your attention. Fixing now.

File: src/main/java/algorithms/curated170/medium/TheMaze.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            int nextX = x + dirx;
+            int nextY = y + diry;
+
+            while (canPass(nextX, nextY)
+                    && !visited[nextX][nextY]) {
+                int prx = nextX;
+                int pry = nextY;
+                nextX += dirx;
+                nextY += diry;
+                result = result || !canPass(nextX, nextY) && dps(prx, pry);
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/TheMaze.java:51]: I think using parenthesis here for the operators would make the code more readable.
As a note for those who do not know about the precedence of the operators. 
image
Source: http://www.math.uaa.alaska.edu/~afkjm/cs110/handouts/IfStatements.pdf

Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/TheMaze.java]: Nice idea to add this table.

File: src/main/java/algorithms/curated170/medium/TheMaze.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            int nextX = x + dirx;
+            int nextY = y + diry;
+
+            while (canPass(nextX, nextY)
+                    && !visited[nextX][nextY]) {
+                int prx = nextX;
+                int pry = nextY;
+                nextX += dirx;
+                nextY += diry;
+                result = result || (!canPass(nextX, nextY) && dfs(prx, pry));
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/TheMaze.java:51]: Let's say that we have went left from some tile. This loop implies that we will also traverse the same from left to right until we see that the right tile is already visited. Isn't this to some degree duplicate unneeded calculation or am I seeing it wrong?

Comment (altayhunoglu) [reply src/main/java/algorithms/curated170/medium/TheMaze.java:51]: It is duplicated but I am not sure if it is unnecessary.

Somehow we have to distinguish the above-mentioned case and the case in Example-2.

It is one of the alternative ways.

We can change it when we find a more efficient way.

Comment (ErdemT09) [reply src/main/java/algorithms/curated170/medium/TheMaze.java:51]: We can perhaps put a simple statement like "Don't go right" to the method call.  We can wrap this up in an integer or a boolean value.

**Comment by altayhunoglu** [comment]
I will also add BFS solution. Maybe it is much more appropriate as it can prevent dead-ends more efficiently.

**Comment by altayhunoglu** [comment]
[Adds BFS, refactor DFS.](https://github.com/spiralgo/algorithms/pull/282/commits/b57f4932d61253bd64e1339836a94774c85a5b65)

**Comment by altayhunoglu** [comment]
[
Adds our first unit test for The Maze - DFS solution.](https://github.com/spiralgo/algorithms/pull/282/commits/b201c6c1455b98f4d3a659f9f639788fff150311)

**Comment by ErdemT09** [comment]
Here is what I meant:
```java
public boolean dfs(int x, int y, int arrivedFrom) {

        if (x == destination[0] && y == destination[1]) {
            return true;
        }

        visited[x][y] = true;
        boolean result = false;

           for (int i = 0; i<4; i++){
               
            if((i == 0 && arrivedFrom == 3)
            ||(i == 1 && arrivedFrom == 2)
            || (i == 2 && arrivedFrom == 1)
            || (i == 3 && arrivedFrom == 0))
            {
                continue;
            }
            int dirx = directions[i][0];
            int diry = directions[i][1];
            int nextX = x + dirx;
            int nextY = y + diry;

            while (canPass(nextX, nextY)
                    && !visited[nextX][nextY]) {
                int prx = nextX;
                int pry = nextY;
                nextX += dirx;
                nextY += diry;
                result = result || !canPass(nextX, nextY) && dfs(prx, pry, i);

            }
       }

        return result;
    }
```

**Comment by altayhunoglu** [comment]
> Here is what I meant:
> 
 
 Nice idea Erdem, thanks.
 I am not sure what is the exact run-time efficiency as it changes in each run:
 
image

But, it is fine to add to be more specific.

**Comment by ErdemT09** [comment]
In the BFS solution, we can perhaps implement the algorithm of not going back to the place we came from by changing the boolean matrix with a `int`, or simply `byte` matrix.

**Comment by ErdemT09** [comment]
I think the solutions are current quite fine. We can add other settings to them in the future if needed.

**Review by ErdemT09 — APPROVED** [review]

**Pull request merged**